src/rust/src/file_reader.rs

```diff
@@ -421,37 +435,44 @@ mod geozero_api {
 // Shared FallibleStreamingIterator implementation
 impl<R: Read, S> FeatureIter<R, S> {
     fn new(
-        mut reader: R,
+        reader: R,
         verify: bool,
-        mut fbs: FgbFeature,
+        fbs: FgbFeature,
         item_filter: Option<Vec<packed_r_tree::SearchResultItem>>,
     ) -> FeatureIter<R, S> {
-        let finished = Self::read_feature_size(&mut fbs, &mut reader);
-        let count = match &item_filter {
+        let mut iter = FeatureIter {
+            reader,
+            verify,
+            fbs,
+            item_filter,
+            count: None,
+            feat_no: 0,
+            cur_pos: 0,
+            state: State::Init,
+            seekable_marker: PhantomData,
+        };
+
+        if iter.read_feature_size() {
+            iter.state = State::Finished
+        } else {
+            iter.state = State::ReadFirstFeatureSize
+        };
+
+        iter.count = match &iter.item_filter {
             Some(list) => Some(list.len()),
             None => {
-                let feat_count = fbs.header().features_count() as usize;
+                let feat_count = iter.fbs.header().features_count() as usize;
                 if feat_count > 0 {
                     Some(feat_count)
-                } else if finished {
+                } else if iter.state == State::Finished {
                     Some(0)
                 } else {
                     None
                 }
             }
         };
 
-        FeatureIter {
-            reader,
-            verify,
-            fbs,
-            item_filter,
-            count,
-            feat_no: 0,
-            cur_pos: 4,
```

Comment:
We used to initialize the `cur_pos: 4` here to signify that we've already read the first feature length. This was necessary to correctly compute the offset if we were to seek to a subsequent feature.